Create main Karpenter concepts doc (#777)
awslabs/karpenter: website/content/en/docs/concepts/_index.md

---
title: "Concepts"
linkTitle: "Concepts"
weight: 35
---

By adding Karpenter to a Kubernetes cluster you can dramatically improve the efficiency and cost of running workloads on that cluster.

> Instead of introducing lots of new concepts and configuration, Karpenter is tightly integrated with Kubernetes features to make sure that the right types and amounts of compute resources are available to pods as they are needed.

Reviewer: Consider dropping `Instead of introducing lots of new concepts and configuration`. I think it's as strong without, and it tends to be more crisp to say what something is, rather than what it isn't.

Karpenter carries out its duties by:

* Watching for pods that the Kubernetes scheduler has marked as unschedulable
* Evaluating the requirements (capacity, features, and taints) identified by the unscheduable pods

> * Provisioning nodes that meet the requirements of the unschedulable pods

Reviewer: We repeat "unschedulable" here a bunch, maybe we can just say it once in the first bullet and then the context can be inferred and simplified to"the pods"

* Scheduling the unschedulable pods to run on the new nodes
* Removing the nodes when the nodes are no longer needed

> In many cases, you can configure a simple Karpenter provisioner when it is first installed and not change it again. 

Reviewer: s/simple/unconstrained

> Other times, you might want to continue to tweak the provisioner or even create multiple provisioners for a cluster where you tailor the requirements of different teams to use specific instance types, zones, or computer architectures.

Reviewer: I'd drop "computer" from "computer architectures".

> Likewise, you might want to instruct the provisioner to take advantage of underlying cloud features, such as AWS spot instances.

Reviewer: s/instruct/constrain
s/take advantage of/use

> As someone deploying pod workloads, you can make specific requests for capacity and features you want from the nodes running your pods.

Reviewer: > As someone deploying pod workloads

I'd love to standardize on terms. Perhaps cluster operator, application developer, end user?

I also find the framing a bit odd here. You're not requesting capacity for its own sake, you're trying to run pods. 

I might say something like, "node features and capacity are determined dynamically based on the requirements of your pods".

Reviewer: We might be able to simply omit end users, since they shouldn't interact much with our layer of abstraction.

> If the Kubernetes scheduler is unable to meet those requirements, Karpenter is designed to quickly create the best possible nodes to meet those needs.

Reviewer: It's not the kube scheduler failing to meet the requirements, it's the existing capacity.

Also -- I'd simplify "karpenter is designed to" to "karpenter quickly ___"

## Karpenter concepts

> The following are some key concepts you should understand about Karpenter.

Reviewer: Consider omitting this small section entirely. It's covered by the header and context of the page.

> ### Karpenter provisioners

Reviewer: Thoughts on being more judicious about putting karpenter throughout the docs? I think we can probably get away with just saying `Provisioners` and avoid repeating.

> A Karpenter provisioner controller runs on a Kubernetes cluster.

Reviewer: "The Karpenter Operator runs ___". Pedantically, it could run anywhere. You could say "Karpenter can be installed onto your Kubernetes cluster's data plane.

Its job is to add nodes to handle unschedulable pods, schedule pods on those nodes, and remove the nodes when they are not needed.
The controller's work is configured through [Provisioner CRDs](/docs/provisioner-crd/).

> Rather than take over the process of provisioning nodes and scheduling pods, the provisioner controller integrates with the Kubernetes scheduler, cluster autoscaler, and related Kubernetes features to make sure that pods are properly and efficiently deployed.

Reviewer: There are a couple inaccuracies here.
>Rather than take over the process of provisioning nodes and scheduling pods, 

I'd remove this rather statement as it may not always be true for this project ;)

>the provisioner controller integrates with the Kubernetes scheduler, cluster autoscaler,

We don't integrate with the cluster autoscaler. They don't play well together unless carefully configured.

>and related Kubernetes features to make sure that pods are properly and efficiently deployed.
 
This reads as a non sequitur to me

> It also works through Kubernetes to request needed assets from the cloud provider.

Reviewer: This implies we're requesting capacity through something like ACK or CAPI, which isn't true. We call the cloud provider directly (not through kubernetes interfaces/crds).

> Here are some attributes of the Karpenter provisioner:

Reviewer: Attributes makes me think of provisioner parameters like `provisioner.spec.requirements`.

Reviewer: In my mind, it's something like:

Concepts:
- provisioner(s)
- well known labels
- provider-specific parameters 
- resource requests
- node affinity (w/ node selector syntactic sugar)
- tolerations
- topology spread constraints
- persistent volumes 

AWS-specific Concepts:
- instance role
- launch template
- security groups
- subnets

Behaviors:
- batching
- scheduling
- binpacking
- termination
- expiration
- emptiness

* **Unschedulable pods**: Karpenter only looks at pods that the Kubernetes scheduler identifies as unschedulable.

> If the Kubernetes scheduler is able to schedule a pod, the Karpenter provisioner takes no action on it.

Reviewer: More crisply, Karpenter only attempts to provision pods that has a status condition `Unschedulable=True`, which the kube scheduler sets when it fails to schedule the pod to existing capacity.

> * **Provisioner CRD**: Provisioner CRDs define how the provisioner controller behaves.

Reviewer: Worth stating that each provisioner manages a distinct set of nodes.

Each CRD contains constraints that impact the nodes that can be provisioned and attributes of those nodes (such timers for removing nodes).
See [Provisioner CRD](/docs/provisioner-crd/) for a description of settings.

> * **Well-known labels**: The CRD can use well-known Kubernetes labels to allow pods to request only certain instance types, computer architectures, operating systems, or other attributes when creating nodes.

Reviewer: Worth linking to https://kubernetes.io/docs/reference/labels-annotations-taints/

> * **Time to live**: A provisioner can also include time-to-live values to indicate when nodes should be decommissioned after a set amount of time from when it was created or after it becomes empty of deployed pods.

Reviewer: This one feels a bit in the weeds to me. We currently only have two controls, but this will balloon over time beyond TTLs. Worth mentioning limits (not yet implemented) if anything at all.

Author: I've added a deprovisioning task section to the new outline. I'll just give a brief mention here and point to the new task.

* **Multiple provisioner CRDs**: An important feature of Karpenter is that it allows multiple provisioner CRDs on the same cluster.
One possible result is that you could configure different teams on the same cluster to run on completely separate capacity.

> One team could run on GPU nodes while another uses storage nodes, with each having a different set of defaults.

Reviewer: It's important to note that this can be handled by one provisioner as well. The reason you might want two is if you want to isolate the nodes (e.g. billing) or if you want different defaults (i.e. no gpus for this team) or behaviors (i.e. expiration/scale down).

This approach could some day lead Karpenter to allow provisioners with different cloud providers on the same cluster or a control plane on one cloud and data plane on another.

> ### Karpenter scheduler

Reviewer: see above re: repeating "Karpenter". Consider:
### Scheduling

Besides creating nodes, Karpenter also works with the Kubernetes schedular to schedule pods.
To avoid conflicts, Karpenter only schedules pods that the Kubernetes scheduler has marked unschedulable.

> When Karpenter finds unschedulable pods, it provisions the needed nodes and immediately binds the previously unschedulable pods to the new nodes.

Reviewer: Karpenter optimistically creates the Node object and binds the pod. This stateless approach avoids race conditions and improves performance. If something is wrong with the launched node, Kubernetes will automatically migrate the pods to a new node.

> Once Karpenter brings up a node, that node is available for the Kubernetes scheduler to schedule pods on it as well.

Reviewer: This is useful if there's additional room in the node due to imperfect packing shape or as workloads finish over time.

### Cloud provider
Karpenter makes requests to provision new nodes to the associated cloud provider.
The first supported cloud provider is AWS, although Karpenter is designed to work other cloud providers as well.
Separating Kubernetes and AWS-specific settings allows Karpenter a clean path to integrating with other cloud providers.

While using Kubernetes well-known labels, the provisioner can set values that are specific to the cloud provider.
So, for example, to include a certain instance type, you could use the Kubernetes label `node.kubernetes.io/instance-type`, but set its value to an AWS instance type (such as `m5.large` or `m5.2xlarge`).

> Other cloud assets you can identify include security groups and capacity type.

Reviewer: You can't specify security groups via well known labels.

"Cloud providers may choose to implement support for additional vendor specific labels like `node.k8s.aws/capacity-type`"

Choosing spot instances is one feature that is specific to AWS.
See [AWS labels](/docs/cloud-providers/aws/) for details.

### Kubernetes cluster autoscaler
Like Karpenter, [Cluster Autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler) is
designed to add nodes when requests come in to run pods that cannot be met by current capacity.
Cluster autoscaler is part of the Kubenetes project, with implementations by most major Kubernetes cloud providers.
By taking a fresh look at provisioning, Karpenter offers the following improvements over the older cluster autoscaler:

> * **Tuned to manage AWS EC2 instances**: 

Reviewer: Designed to handle the full flexibility of the cloud.

> At the time cluster autoscaler was created, cloud providers, such as GCE, only had about a half dozen instance types.

Reviewer: This is a very difficult statement to defend. Consider removing, since the second statement covers this.

> Cluster autoscaler was not built with the flexibility to handle hundreds of instance types or request EC2 spot instances.

Reviewer: flexibility of hundreds of instance types, zones, and purchase options.

Karpenter has the ability to efficiently address the full range of instance types available through AWS.

* **Adds scheduling**: Cluster autoscaler doesn’t bind pods to the nodes it creates.
A node that Karpenter creates knows to run the bound pods immediately when the node comes up.
It doesn't have to wait for the scheduler or for the node to become ready.
It can start pulling the container images it needs immediately.

> This alone can save an extra 15 seconds of performance.

Reviewer: This statement is hard to defend without data. Consider saying "seconds".

> ### Constraints

Reviewer: Didn't have time to make it through this section or below.

Author: Okay. I've responded to everything else to this point (or asked for further clarification).

* Provisioning nodes that meet the requirements of the pods
* Scheduling the pods to run on the new nodes
* Removing the nodes when the nodes are no longer needed

In many cases, you can configure an unconstrained Karpenter provisioner when it is first installed and not change it again. 
Other times, you might want to continue to tweak the provisioner or even create multiple provisioners for a cluster where you tailor the requirements of different teams to use specific instance types, zones, or architectures.
Likewise, you might want to constrain the provisioner to use underlying cloud features, such as AWS spot instances.

Application developers can make specific requests for capacity and features you want from the nodes running your pods.
Karpenter is designed to quickly create the best possible nodes to meet those needs.

## Provisioners

Karpenter can be installed onto your Kubernetes cluster's data plane.
Its job is to add nodes to handle unschedulable pods, schedule pods on those nodes, and remove the nodes when they are not needed.
Karpenter's provisioning work is configured through [Provisioner CRDs](/docs/provisioner-crd/).

> Here are some things to know about the Karpenter provisioner:

Author: Okay. I reorganized the content to tie each section to cluster operator and application developer roles.

* **Unschedulable pods**: Karpenter only attempts to provision pods that has a status condition Unschedulable=True, which the kube scheduler sets when it fails to schedule the pod to existing capacity.

* **Provisioner CRD**: Provisioner CRDs define how the provisioner controller behaves.
Each provisioner manages a distinct set of nodes.

> Each CRD contains constraints that impact the nodes that can be provisioned and attributes of those nodes (such timers for removing nodes).

Author: Done.

See [Provisioner CRD](/docs/provisioner-crd/) for a description of settings.
* **Well-known labels**: The CRD can use well-known Kubernetes labels to allow pods to request only certain instance types, architectures, operating systems, or other attributes when creating nodes.
See [Well-Known Labels, Annotations and Taints](https://kubernetes.io/docs/reference/labels-annotations-taints/) for details.

> Keep in mind that only a subset of these labels are supported in Karpenter, as described in [Provisioning](/docs/concepts/provisioner.md).

Reviewer 2: There's also Karpenter specific well known labels (currently just capacity type though)

* **Time to live**: A provisioner can also include time-to-live values to indicate when nodes should be decommissioned after a set amount of time from when it was created or after it becomes empty of deployed pods.

> * **Multiple provisioner CRDs**: An important feature of Karpenter is that it allows multiple provisioner CRDs on the same cluster.

Author: Done.

One possible result is that you could configure different teams on the same cluster to run on completely separate capacity.
One team could run on GPU nodes while another uses storage nodes, with each having a different set of defaults.
Although you can address the needs of multiple teams in one CRD in some ways, having separate ones lets you isolate nodes for billing, use different defaults (such as no GPUs for a team), or use different expiration and scale-down values.

> Having multiple CRDs could some day lead Karpenter to allow provisioners with different cloud providers on the same cluster or a control plane on one cloud and data plane on another.

Author: Done.

## Scheduling

Besides creating nodes, Karpenter also schedules pods.
To avoid conflicts, Karpenter only schedules pods that the Kubernetes scheduler has marked unschedulable.
Karpenter optimistically creates the Node object and binds the pod.

> This stateless approach avoids race conditions and improves performance.

Author: Done.

If something is wrong with the launched node, Kubernetes will automatically migrate the pods to a new node.

Once Karpenter brings up a node, that node is available for the Kubernetes scheduler to schedule pods on it as well.
This is useful if there is additional room in the node due to imperfect packing shape or as workloads finish over time.

## Cloud provider
Karpenter makes requests to provision new nodes to the associated cloud provider.

> The first supported cloud provider is AWS, although Karpenter is designed to work other cloud providers as well.

Reviewer: I might leave this out, as it may not age well. I think it's clear that only AWS is supported by the existence of the cloud provider section.

Separating Kubernetes and AWS-specific settings allows Karpenter a clean path to integrating with other cloud providers.

While using Kubernetes well-known labels, the provisioner can set some values that are specific to the cloud provider.
So, for example, to include a certain instance type, you could use the Kubernetes label `node.kubernetes.io/instance-type`, but set its value to an AWS instance type (such as `m5.large` or `m5.2xlarge`).

> Choosing spot instances is one feature that is specific to AWS.

Reviewer: This isn't actually accurate. I'd remove it. We're trying to generalize this concept, but just haven't done it yet.

Author: I'll not that it's AWS for now but could be generalized in the future.

> Cloud providers may choose to implement support for additional vendor specific labels like `node.k8s.aws/capacity-type`.

Reviewer: ```suggestion
Further, cloud providers may choose to implement support for additional vendor specific labels like `node.k8s.aws/capacity-type`.
```

Author: Done.

See [AWS labels](/docs/cloud-providers/aws/) for details.

> ## Kubernetes cluster autoscaler

Reviewer: Let's push this section to the bottom.

Like Karpenter, [Cluster Autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler) is
designed to add nodes when requests come in to run pods that cannot be met by current capacity.
Cluster autoscaler is part of the Kubenetes project, with implementations by most major Kubernetes cloud providers.
By taking a fresh look at provisioning, Karpenter offers the following improvements over the older cluster autoscaler:

* **Designed to handle the full flexibility of the cloud**:
Cluster autoscaler was not built with the flexibility to handle hundreds of instance types, zones, and purchase options.
Karpenter has the ability to efficiently address the full range of instance types available through AWS.

* **Adds scheduling**: Cluster autoscaler doesn’t bind pods to the nodes it creates.
A node that Karpenter creates knows to run the bound pods immediately when the node comes up.
I doesn't have to wait for the scheduler or for the node to become ready.
It can start pulling the container images it needs immediately.
This can save seconds of performance.

> ## Constraints

Reviewer: Let's sit down and talk this one through. I started reviewing and then realized it might be easier over voice..

Understanding the concept of layered constraints is key to using Karpenter.

> With no constraints defined in provisioners and none requested from pods being deployed, Karpenter on AWS is free to choose from the entire universe of features available from EC2.

Reviewer: Thoughts on how we can stay vendor neutral here?

> With no constraints defined in provisioners and none requested from pods being deployed, Karpenter on AWS is free to choose from the entire universe of features available from EC2.

Reviewer: ```suggestion
With no constraints defined in provisioners and none specified from pods being deployed, Karpenter is free to choose from the entire universe of features available from your cloud provider.
```

Nodes can be created using any instance type and run in any zones.

> On-demand or spot instances can be used.

Reviewer: Thoughts on omitting this one for aforementioned neutrality reaosns.

> The effects of layering constraints are applied when a pod makes requests for compute resources that cannot be met by current capacity.

Reviewer: I'd drop this since we already cover when karpenter kicks in above (unschedulable)

> A pod can request *affinity* (such as to run in a particular zone or instance type) or a topology spread (to require that a set of pods be run across multiple nodes).

Reviewer: ```suggestion
A pod can specify `nodeAffinity` (such as to run in a particular zone or instance type) or `topologySpreadConstraints` (to require that a set of pods be spread across multiple zones or nodes).
```

The pod can ask to run only on nodes with a particular label or that have a certain amount of available memory or particular architecture type.

The Kubernetes scheduler then tries to match those constraints with available nodes.
If the node is unschedulable, facilities like the cluster autoscaler or Karpenter come in and try to create compute resources to match each pod's needs.
When Karpenter tries to provision a node that can schedule an unschedulable pode, it compares its own constraints when choosing the node to create.
The Karpenter provisoners configured for the cluster will look to best match the request, using mostly the same well-known labels and affinity features that the pod uses.

For details about how to set constraints when configuring a Karpenter provider, see [Karpenter provisioner and scheduler](/docs/concepts/provisioner.md) for details.
For more on how, as a developer, you can add constraints to your pod spec deployment, see [Karpenter pod developer concepts](/docs/concepts/running-pods.md) for details.

## Deprovisioning nodes

> Karpenter not only manages the creation of nodes to add cluster capacity, but also will delete nodes when that capacity is no longer needed.

Reviewer: ```suggestion
Karpenter will delete node when they are no longer needed.
```

> There are several things you should know about how Karpenter deprovisions nodes:

Reviewer: This line is self evident, imo. Thoughts on cutting it?

* **Node Expiry**: If a node expiry time-to-live value (`ttlSecondsUntilExpired`) is reached, that node is drained of pods and deleted (even if it is still running workloads).
* **Empty nodes**: When the last workload pod running on a Karpenter-managed node is gone, a clock starts keeping track of how long the node is empty.

> Once that "node empty" time-to-live (`ttlSecondsAfterEmpty`) is reached, that node is deleted.

Reviewer: ```suggestion
Once that "node empty" time-to-live (`ttlSecondsAfterEmpty`) is reached, finalization is triggered.
```

> * **Finalizer**: Karpenter places a finalizer bit on the nodes it creates.

Reviewer: Thoughts putting this first and explaining cordon/drain once, since it's shared between all flows.

> When a request comes in to delete one of those nodes (from an expired TTL or maybe an explicit `kubectl delete node`), Karpenter will drain all the pods and then terminate the EC2 instance.

Reviewer: ```suggestion
When a request comes in to delete one of those nodes (e.g. expiration or even an explicit `kubectl delete node`), Karpenter will drain all the pods, call delete on the cloud provider, and finally delete the node.
```

Karpenter handles all clean-up work needed to properly delete the node.

For more details on how Karpenter deletes nodes, see [Deleting nodes with Karpenter](/docs/concepts/delete-nodes.md) for details.

> ## Karpenter workflows

Reviewer 2: Isn't the reconciliation workflow missing? At least that's what I think the workflow is called that creates nodes in response to unschedulable pods

To further understand Karpenter concepts and how Karpenter works, refer to concept documentation on the following workflows:

> * **Setting up Karpenter**: To get started with Karpenter, an operator needs to deploy Karpenter on a running cluster and provide it with the necessary permissions (IAM Roles for Service Accounts for AWS, in this example) to create and delete nodes on the cloud provider from within Kubernetes.

Reviewer: ```suggestion
* **Setting up Karpenter**: To get started with Karpenter, an operator needs to deploy Karpenter on a running cluster and provide it with the necessary cloud provider credentials (e.g. IAM Roles for Service Accounts for AWS) to create and delete nodes.
```

> * **Configuring Provisioners**: Configuring one or more provisioners on a cluster allows an operator to define the constraints by which nodes are created and deleted on the assigned cluster.

Reviewer: ```suggestion
* **Configuring Provisioners**: Configuring one or more provisioners on a cluster allows an operator to define constraints for nodes that are created, and controls for when to delete them.
```

> * **Deploying pods**: A developer that is configuring pods can request that features like taints and affinity be met by the nodes that are created to run those pods.

Reviewer: ```suggestion
* **Deploying pods**: A developer that is configuring pods can specify `tolerations` or `nodeAffinity` that constrain the nodes that are created to run those pods.
```

Karpenter is made to match those requests to the best available nodes, then schedule the pods on those nodes.

> * **Deleting nodes**: Karpenter will drain pods and delete nodes based on direct requests or timer expirations.

Reviewer: ```suggestion
* **Deleting nodes**: Karpenter has a standardized node draining and deletion workflow, which is triggered by direct requests, emptiness, or expiration.
```

Although some use cases are addressed witha single provisioner for multiple teams, multiple provisioners are useful to isolate nodes for billing, use different defaults (such as no GPUs for a team), or use different deprovisioning settings.

In the future, multiple provisioners would enable to provision nodes from different cloud providers on the same cluster or even a control plane on one cloud and data plane on another.

### Deprovisioning nodes

Karpenter not only manages the creation of nodes to add cluster capacity, but also will delete nodes when that capacity is no longer needed.
There are several things you should know about how Karpenter deprovisions nodes:

* **Node Expiry**: If a node expiry time-to-live value (`ttlSecondsUntilExpired`) is reached, that node is drained of pods and deleted (even if it is still running workloads).
* **Empty nodes**: When the last workload pod running on a Karpenter-managed node is gone, a clock starts keeping track of how long the node is empty.
Once that "node empty" time-to-live (`ttlSecondsAfterEmpty`) is reached, that node is deleted.
* **Finalizer**: Karpenter places a finalizer bit on each node it creates.

> When a request comes in to delete one of those nodes (from an expired TTL or maybe an explicit `kubectl delete node`), Karpenter will drain all the pods and then terminate the EC2 instance.

Reviewer: ```suggestion
When a request comes in to delete one of those nodes (e.g., a TTL or a manual `kubectl delete node`), Karpenter will drain all the pods, terminate the EC2 instance, and delete the node object.
```

Karpenter handles all clean-up work needed to properly delete the node.

For more details on how Karpenter deletes nodes, see [Deleting nodes with Karpenter](/docs/tasks/delete-nodes.md) for details.

### Upgrading nodes

> A straight-forward way to upgrade nodes that are managed by Karpenter is to add new upgraded nodes, move pod workloads from the old nodes to the new ones, and delete the old nodes.
> Another way to do upgrades is to set `ttlSecondsUntilExpired` values so old nodes are deleted after a set period of time, so they can then be replaced with newer nodes.
> For details on upgrading nodes with Karpenter, see [Upgrading nodes with Karpenter](/docs/tasks/upgrade-nodes.md) for details.

Reviewer: In practice, the first example is likely challenging to orchestrate for users.
```suggestion
A straight-forward way to upgrade nodes is to set `ttlSecondsUntilExpired`. Nodes will be terminated after a set period of time and will be replaced with newer nodes.
For details on upgrading nodes with Karpenter, see [Upgrading nodes with Karpenter](/docs/tasks/upgrade-nodes.md) for details.
```

> ### Cluster operator concepts

Reviewer: This h3 reads a bit odd to me since we're already in the h2 Cluster Operator. Perhaps simpler to promote each of these h4s to h3.
```suggestion
```

Understanding the following concepts will help you in carrying out the tasks just described.

#### Constraints

The concept of layered constraints is key to using Karpenter.

> With no constraints defined in provisioners and none requested from pods being deployed, Karpenter on AWS is free to choose from the entire universe of features available from EC2.

Reviewer: ```suggestion
With no constraints defined in provisioners and none requested from pods being deployed, Karpenter chooses from the entire universe of features available from your cloud provider.
```

Nodes can be created using any instance type and run in any zones.

> On-demand or spot instances can be used.

Reviewer: We actually default to OD for AWS. Spot is opt in.
```suggestion
```

> Most of the same constraints a Karpenter cluster operator can set up on a provisioner can be used by an application developer in a podspec to request particular features from a node.

Reviewer: ```suggestion
Constraints defined in a provisioner by the cluster operator can be tightened by an application developer by defining additional scheduling constraints in their pod spec.
```

Refer to the description of Karpenter constraints in the Application Developer section below for details.

#### Scheduling

> To avoid conflicts, Karpenter only schedules pods that the Kubernetes scheduler has marked unschedulable.

Reviewer: ```suggestion
Karpenter schedules pods that the Kubernetes scheduler has marked unschedulable.
```

> Karpenter optimistically creates the Node object and binds the pod.

Reviewer: ```suggestion
After solving scheduling constraints and launching capacity, Karpenter optimistically creates the Node object and binds the pod.
```

This stateless approach helps to avoid race conditions and improves performance.
If something is wrong with the launched node, Kubernetes will automatically migrate the pods to a new node.

Once Karpenter brings up a node, that node is available for the Kubernetes scheduler to schedule pods on it as well.
This is useful if there is additional room in the node due to imperfect packing shape or because workloads finish over time.

#### Cloud provider
Karpenter makes requests to provision new nodes to the associated cloud provider.

> The first supported cloud provider is AWS, although Karpenter is designed to eventually work with other cloud providers as well.

Reviewer: ```suggestion
The first supported cloud provider is AWS, although Karpenter is designed to with other cloud providers.
```

Separating Kubernetes and AWS-specific settings allows Karpenter a clean path to integrating with other cloud providers.

While using Kubernetes well-known labels, the provisioner can set some values that are specific to the cloud provider.
So, for example, to include a certain instance type, you could use the Kubernetes label `node.kubernetes.io/instance-type`, but set its value to an AWS instance type (such as `m5.large` or `m5.2xlarge`).

> Choosing spot instances is one feature that is currently specific to AWS, but one that could be generalized to other providers.
> Furhter, cloud providers may choose to implement support for additional vendor specific labels like `node.k8s.aws/capacity-type`.
> See [AWS labels](/docs/cloud-providers/aws/) for details.

Reviewer: The project has evolved since we spoke :). This is now generically available as `karpenter.sh/capacity-type`
```suggestion
```

#### Kubernetes cluster autoscaler
Like Karpenter, [Kubernetes Cluster Autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler) is
designed to add nodes when requests come in to run pods that cannot be met by current capacity.
Cluster autoscaler is part of the Kubenetes project, with implementations by most major Kubernetes cloud providers.
By taking a fresh look at provisioning, Karpenter offers the following improvements:

* **Designed to handle the full flexibility of the cloud**:
Karpenter has the ability to efficiently address the full range of instance types available through AWS.
Cluster autoscaler was not originally built with the flexibility to handle hundreds of instance types, zones, and purchase options.

> 

Reviewer: ```suggestion

**Group-less node provisioning**:
Karpenter manages each instance directly, without use of additional orchestration mechanisms like node groups. This enables it to retry in milliseconds instead of minutes when capacity is unavailable. It also allows Karpenter to leverage diverse instance types, availability zones, and purchase options without creation of hundreds of node groups.

```

> * **Adds scheduling**: Cluster autoscaler doesn’t bind pods to the nodes it creates.

Reviewer: ```suggestion
* **Scheduling Enforcement**: Cluster autoscaler doesn’t bind pods to the nodes it creates. Instead, it relies on the kube-scheduler to make the same scheduling decision after the node has come online.
```

> A node that Karpenter creates knows to run the bound pods immediately when the node comes up.

Reviewer: ```suggestion
A node launched by Karpenter has its pods bound immediately.
```

> I doesn't have to wait for the scheduler or for the node to become ready.

Reviewer: ```suggestion
The kubelet doesn't have to wait for the scheduler or for the node to become ready.
```

> It can start pulling the container images it needs immediately.

Reviewer: ```suggestion
It can start preparing the container runtime immediately, including pre-pulling the image.
```

> This can save seconds of performance.

Reviewer: ```suggestion
This can shave seconds off of node startup latency.
```

## Application developer

As someone deploying pods that might be evaluated by Karpenter, you should know how to request the properties that your pods need of its compute resources.
Karpenter's job is to efficiently assess and choose compute assets based on requests from pod deployments.
These can include basic Kubernetes features or features that are specific to the cloud provider (such as AWS).

Layered *constraints* are applied when a pod makes requests for compute resources that cannot be met by current capacity.

> A pod can request *affinity* (to run in a particular zone or instance type) or a topology spread (to cause a set of pods be run across multiple nodes).

Reviewer: ```suggestion
A pod can specify `nodeAffinity` (to run in a particular zone or instance type) or `topologySpreadConstraints` (to cause a set of pods be balanced across multiple nodes).
```

> The pod can ask to run only on nodes with a particular label or that have a certain amount of available memory or particular architecture type.

Reviewer: ```suggestion
The pod can specify a `nodeSelector` to run only on nodes with a particular label and  `resource.requests` to ensure that the node has enough available memory.
```

The Kubernetes scheduler tries to match those constraints with available nodes.

> If the pod is unschedulable, facilities like the Cluster Autoscaler or Karpenter come in and try to create compute resources to match each pod's needs.

Reviewer: ```suggestion
If the pod is unschedulable, Karpenter creates compute resources that match its needs.
```

> When Karpenter tries to provision a node that can schedule an unschedulable pode, it compares its own constraints when choosing the node to create.

Reviewer: ```suggestion
When Karpenter tries to provision a node, it analyzes scheduling constraints before choosing the node to create.
```

As long as the requests are not outside of the provisioner's constraints, 

> Karpenter will look to best match the request, comparing mostly the same well-known labels and affinity features that the pod uses.

Reviewer: ```suggestion
Karpenter will look to best match the request, comparing the same well-known labels defined by the pod's scheduling constraints.
```

> Note that if the constraints are such that a match is not possible, the pod will not deploy.

Reviewer: ```suggestion
Note that if the constraints are such that a match is not possible, the pod will remain unscheduled.
```

So, what constraints can you use as an application developer deploying pods that could be managed by Karpenter?

* **Kubernetes**: Kubernetes features that Karpenters supports for scheduling nodes include node affinity based on [persistant volumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#node-affinity) and [nodeSelector](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#nodeselector).
It also supports [PodDisruptionBudget](https://kubernetes.io/docs/tasks/run-application/configure-pdb/) and [topologySpreadConstraints](https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints/).

  From the Kubernetes [Well-Known Labels, Annotations and Taints](https://kubernetes.io/docs/reference/labels-annotations-taints/) page,
you can see a full list of Kubernetes labels, annotations and taints that determine scheduling.
Only a small set of them are implemented in Karpenter, including:

  * **kubernetes.io/arch**: For example, kubernetes.io/arch=amd64

>   * **kubernetes.io/os**: For example, kubernetes.io/os=linux

Reviewer: This was removed since only linux was supported
```suggestion
```

  * **node.kubernetes.io/instance-type**: For example, node.kubernetes.io/instance-type=m3.medium
  * **topology.kubernetes.io/zone**: For example, topology.kubernetes.io/zone=us-east-1c

> * **AWS-specific**: Capacity types (node.k8s.aws/capacity-type) that are specific to AWS in Karpenter include spot and on-demand.
> Karpenter also supports AWS-specific accelerators, including:
> 
>   * **nvidia.com/gpu**: For Nvidia graphics processing units
>   * **amd.com/gpu**: For AMD graphics processing units
>   * **aws.amazon.com/neuron**: For AWS Neuron machine learning applications

Reviewer: None of these are aws-specific anymore.
```suggestion
* **AWS-specific**: Capacity types (node.k8s.aws/capacity-type) that are specific to AWS in Karpenter include spot and on-demand.
Karpenter also supports AWS-specific accelerators, including:

  * **nvidia.com/gpu**: For Nvidia graphics processing units
  * **amd.com/gpu**: For AMD graphics processing units
  * **aws.amazon.com/neuron**: For AWS Neuron machine learning applications
```

Author: Okay. Since all the AWS-specific settings are going, I'm going to remove the whole AWS-specific section and remove the bullet from the Kubernetes section.

> **NOTE**: Don't use podAffinity and podAntiAffinity to schedule pods on the same or different nodes as other pods.
> Kubernetes SIG scalability recommends against these features and Karpenter doesn't support them.

Reviewer: ```suggestion
**NOTE**: Don't use `podAffinity` and `podAntiAffinity` to schedule pods on the same or different nodes as other pods.
Kubernetes SIG scalability recommends against these features and Karpenter doesn't support them. Instead, we recommend `topologySpreadConstraints` to reduce blast radius and `nodeSelectors` and `taints` to implement colocation.
```

Author: Nice. Done.
